core/src/main/java/org/elasticsearch/env/Environment.java

```diff
@@ -138,9 +145,32 @@ public Environment(Settings settings) {
                 dataFiles[i] = PathUtils.get(dataPaths.get(i));
                 dataWithClusterFiles[i] = dataFiles[i].resolve(clusterName.value());
             }
+            if (DEFAULT_PATH_DATA_SETTING.exists(settings)) {
+                final String defaultPathDataValue = DEFAULT_PATH_DATA_SETTING.get(settings);
+                final Set<Path> dataFilesSet = Arrays.stream(dataFiles).collect(Collectors.toSet());
+                final Path defaultPathData = PathUtils.get(defaultPathDataValue);
+                if (dataFilesSet.size() == 1 && dataFilesSet.contains(defaultPathData)) {
+                    // default path data was used to set path data
+                    this.defaultPathData = null;
+                } else if (dataFilesSet.contains(defaultPathData)) {
+                    final String message = String.format(
+                            Locale.ROOT,
+                            "do not include default.path.data [%s] in path.data %s",
+                            defaultPathData,
+                            Arrays.toString(dataFiles));
+                    throw new IllegalStateException(message);
+                } else {
+                    this.defaultPathData = defaultPathData;
+                }
+            } else {
+                defaultPathData = null;
+            }
         } else {
             dataFiles = new Path[]{homeFile.resolve("data")};
             dataWithClusterFiles = new Path[]{homeFile.resolve("data").resolve(clusterName.value())};
+            assert !DEFAULT_PATH_DATA_SETTING.exists(settings)
+                    : "expected default.path.data to be unset but was [" + DEFAULT_PATH_DATA_SETTING.get(settings) + "]";
```

Comment:
This is a case when *nothing* was set not even `default.path.data` (think of starting from an archive distribution). You see, if `path.data` is not set and `default.path.data` is, then `default.path.data` is copied into `path.data` and we are not in this branch of the code, instead it's covered by the branch above us when the condition data paths is not empty.